keps/sig-node/20190717-seccomp-ga.md

```diff
@@ -0,0 +1,378 @@
+
+If _only_ seccomp annotations are specified, copy the values into the corresponding fields. This
+ensures that existing applications continue to enforce seccomp, and prevents the kubelet from
+needing to resolve annotations & fields.
+
+If both seccomp annotations _and_ fields are specified, the values MUST match. This will be enforced
+in API validation.
+
+#### Pod Update
+
+The seccomp fields on a pod are immutable.
+
+The behavior on annotation update is currently ill-defined: the annotation update is allowed, but
+the new value will not be used until the container is restarted. There is no way to tell (from the
+API) what value a container is using.
+
+Therefore, seccomp annotation updates will be ignored. This maintains backwards API compatibility
+(no tightening validation), and makes a small stabilizing change to behavior (new Kubelets will
+ignore the update).
+
+#### PodSecurityPolicy Creation
+
+_Similar policy to [Pod Creation](#pod-creation)_
+
+If no seccomp annotations or fields are specified, no action is necessary.
+
+If _only_ seccomp fields are specified, add the corresponding annotations. This ensures that the
+fields are enforced even in skewed high-availability environments. Since [we
+support](https://kubernetes.io/docs/setup/release/version-skew-policy/) up to 1 minor release of
+version skew between the HA masters, this behavior will be removed in version N+1 (where N is the
+release with the upgraded seccomp support).
+
+If _only_ seccomp annotations are specified, copy the values into the corresponding fields. This
+smooths the migration for a future release that removes annotation support, and frees tools from
+needing to check annotations.
+
+If both seccomp annotations _and_ fields are specified, the values MUST match. This will be enforced
+in API validation.
+
+#### PodSecurityPolicy Update
+
+If the old pod did not specify any seccomp options (fields or annotations), follow the same rules as
+for creation.
+
+If only the fields or only the annotations are changed, take the new value and copy it to both the
+corresponding fields & annotations (even if one was previously unset).
+
+If both the fields _and_ annotations are changed, the new values MUST match. Enforced in API
+validation.
+
+#### PodSecurityPolicy Enforcement
+
+The PodSecurityPolicy admission controller must continue to check the PSP object for annotations if
+the fields are unset, in order to handle upgrade & version skew scenarios.
+
+When setting default profiles, PSP only needs to set the field. The API machinery will handle
+setting the annotation as necessary.
+
+When enforcing allowed profiles, the PSP should check BOTH the annotations & fields. In most cases,
+they should be consistent. On pod update, the seccomp annotations may differ from the fields. In
+that case, the PSP enforcement should check both values as the effective value depends on the node
+version running the pod.
+
+#### PodTemplates
```

Comment:
consider how cluster operators will know they have things that need updating. a metric that fires on any podtemplate object with seccomp annotations, and (once we drop pod field -> annotation defaulting) any pod with a seccomp annotation might be helpful for tracking this

Reply:
Good idea. It could be any combination of:
- Metric counter
- Log line
- Event (on the controller)
- Warning annotation (on the controller or created pod)

A metric seems too hidden, if you don't know to look for it. I like the event, except it will disappear if you don't notice it right away. I'm leaning towards an annotation on both the controller object and the pod:
```
warning.kubernetes.io/seccomp: "Seccomp set through annotations. Support for setting seccomp through annotations will be dropped in v1.X"
```

We may also want to consider forbidding setting the seccomp annotation after we drop support, so that pods aren't being run `unconfined` unexpectedly. That would be more likely to break users, but follows the secure failure principle.